apis/kms/v1alpha1/autokeyconfig_reference.go

```diff
@@ -0,0 +1,168 @@
+	if r.External != "" && r.Name != "" {
+		return "", fmt.Errorf("cannot specify both name and external on %s reference", KMSAutokeyConfigGVK.Kind)
+	}
+	// From given External
+	if r.External != "" {
+		if _, err := ParseKMSAutokeyConfigExternal(r.External); err != nil {
+			return "", err
+		}
+		return r.External, nil
+	}
+
+	// From the Config Connector object
+	if r.Namespace == "" {
+		r.Namespace = otherNamespace
+	}
+	key := types.NamespacedName{Name: r.Name, Namespace: r.Namespace}
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(KMSAutokeyConfigGVK)
+	if err := reader.Get(ctx, key, u); err != nil {
+		if apierrors.IsNotFound(err) {
+			return "", k8s.NewReferenceNotFoundError(u.GroupVersionKind(), key)
+		}
+		return "", fmt.Errorf("reading referenced %s %s: %w", KMSAutokeyConfigGVK, key, err)
+	}
+	// Get external from status.externalRef. This is the most trustworthy place.
+	actualExternalRef, _, err := unstructured.NestedString(u.Object, "status", "externalRef")
+	if err != nil {
+		return "", fmt.Errorf("reading status.externalRef: %w", err)
+	}
+	if actualExternalRef == "" {
+		return "", k8s.NewReferenceNotReadyError(u.GroupVersionKind(), key)
+	}
+	r.External = actualExternalRef
+	return r.External, nil
+}
+
+// New builds a KMSAutokeyConfigRef from the Config Connector KMSAutokeyConfig object.
+func NewKMSAutokeyConfigRef(ctx context.Context, reader client.Reader, obj *KMSAutokeyConfig) (*KMSAutokeyConfigRef, error) {
+	id := &KMSAutokeyConfigRef{}
+
+	// Get Parent
+	folderRef, err := refsv1beta1.ResolveFolder(ctx, reader, obj, obj.Spec.FolderRef)
+	if err != nil {
+		return nil, err
+	}
+	folderID := folderRef.FolderID
+	if folderID == "" {
+		return nil, fmt.Errorf("cannot resolve project")
+	}
+	id.parent = &KMSAutokeyConfigParent{FolderID: folderID}
+
+	// Use approved External
+	externalRef := valueOf(obj.Status.ExternalRef)
+	if externalRef == "" {
+		id.External = AsKMSAutokeyConfigExternal(id.parent)
+		return id, nil
+	}
+
+	// Validate desired with actual
+	actualParent, err := ParseKMSAutokeyConfigExternal(externalRef)
+	if err != nil {
+		return nil, err
+	}
+	if actualParent.FolderID != folderID {
+		return nil, fmt.Errorf("spec.folderRef changed, expect %s, got %s", actualParent.FolderID, folderID)
+	}
+	id.External = externalRef
+	id.parent = &KMSAutokeyConfigParent{FolderID: folderID}
+	return id, nil
+}
+
+func (r *KMSAutokeyConfigRef) Parent() (*KMSAutokeyConfigParent, error) {
+	if r.parent != nil {
+		return r.parent, nil
+	}
+	if r.External != "" {
+		parent, err := ParseKMSAutokeyConfigExternal(r.External)
+		if err != nil {
+			return nil, err
+		}
+		return parent, nil
+	}
+	return nil, fmt.Errorf("KMSAutokeyConfigRef not initialized from `NewKMSAutokeyConfigRef` or `NormalizedExternal`")
+}
+
+type KMSAutokeyConfigParent struct {
+	FolderID string
+	Location string
```

Comment:
Suggest removing the location because the KMSAutokeyConfig does not have location in its parent URL.